File: web/src/lib/white-balance.ts
import type { WhiteBalanceValues } from "../types";

const D65_CCT = 6504;
const D65_XY = [0.3127, 0.329] as const;
const TINT_DUV_PER_STEP = -0.0005;

const PLANCK_U = [
  0.20042808258946535, 0.021670127838210658, 0.004941279052065094,
  -0.0008090840314764964, -0.00011318638614341891, 0.0001065014881380249,
  -0.000029459270200989892, -0.000002902435510008949, 0.00000572817056676121,
  -0.0000012719911614227974, -0.0000002565272318851006,
] as const;
const PLANCK_V = [
  0.31033342890275567, 0.029893630533631914, -0.0044257853388274,
  -0.0013099604266659877, 0.0008113359906466061, -0.00018080749013704164,
  -0.000019086191775832478, 0.00003585686520700713, -0.000013826413811352124,
  -0.000000675784317713839, 0.0000017403410985881575,
] as const;
const PLANCK_TANGENT_U = [
  0.5869182180956649, 0.2893814826648264, -0.019287578494842157,
  -0.03623096990279516, 0.007623792870567206, 0.0030520940866540953,
  -0.0010403026324977203, -0.00018529622788312033, 0.00008666191184300517,
  0.000008512881065854595, -0.000003949015108301125,
] as const;
const PLANCK_TANGENT_V = [
  0.8096462223108812, -0.20977465709442028, -0.06490896492751604,
  0.016340113139321086, 0.008825176667254188, -0.002203072113364394,
  -0.0009946316564358089, 0.0003055469538525697, 0.00007673679846535353,
  -0.000026349176456116718, -0.0000031870682560505213,
] as const;

const BRADFORD = [
  [0.8951, 0.2664, -0.1614],
  [-0.7502, 1.7135, 0.0367],
  [0.0389, -0.0685, 1.0296],
] as const;
const BRADFORD_INVERSE = [
  [0.9869929054667123, -0.14705425642099013, 0.15996265166373125],
  [0.4323052697233945, 0.5183602715367776, 0.049291228212855594],
  [-0.008528664575177328, 0.04004282165408487, 0.96848669578755],
] as const;
const PROPHOTO_D65_TO_XYZ = [
  [0.755603256421359, 0.11278492113801272, 0.08208189343532289],
  [0.2683379250450128, 0.7151267706955571, 0.016535310335320977],
  [0.003910020350449157, -0.012918708286404542, 1.0978387753557597],
] as const;
const XYZ_TO_PROPHOTO_D65 = [
  [1.4032152671158387, -0.22314009767162846, -0.10155304928343209],
  [-0.5262714954210982, 1.4816610915442805, 0.017031312123848466],
  [-0.011190484846281556, 0.018230026296336695, 0.9114426630797009],
] as const;

type MatrixRow = readonly [number, number, number];
type Matrix = readonly [MatrixRow, MatrixRow, MatrixRow];

/** Builds Studio-compatible relative Bradford adaptation in linear ProPhoto D65. */
export function whiteBalanceMatrix({
  temperature,
  tint,
}: WhiteBalanceValues): Float32Array {
  if (
    !Number.isFinite(temperature) ||
    !Number.isFinite(tint) ||
    temperature < -100 ||
    temperature > 100 ||
    tint < -100 ||
    tint > 100
  ) {
    throw new Error(
      "White balance temperature and tint must be within -100..=100.",
    );
  }
  if (temperature === 0 && tint === 0) {
    return new Float32Array([1, 0, 0, 0, 1, 0, 0, 0, 1]);
  }

  const targetMired = 1_000_000 / D65_CCT + temperature;
  const cct = 1_000_000 / targetMired;
  const uv = planckUvWithDuv(cct, tint * TINT_DUV_PER_STEP);
  const denominator = 2 * uv[0] - 8 * uv[1] + 4;
  const targetXy: [number, number] = [
    (3 * uv[0]) / denominator,
    (2 * uv[1]) / denominator,
  ];
  const adaptation = chromaticAdaptation(xyToXyz(D65_XY), xyToXyz(targetXy));
  return new Float32Array(
    multiplyMatrices(
      multiplyMatrices(XYZ_TO_PROPHOTO_D65, adaptation),
      PROPHOTO_D65_TO_XYZ,
    ).flat(),
  );
}

/** Writes a row-major 3×3 matrix as three WGSL-aligned vec4 rows. */
export function writeWhiteBalanceUniform(
  view: DataView,
  offset: number,
  matrix: Float32Array,
): void {
  for (let row = 0; row < 3; row += 1) {
    for (let column = 0; column < 3; column += 1) {
      view.setFloat32(
        offset + row * 16 + column * 4,
        matrix[row * 3 + column],
        true,
      );
    }
  }
}

function planckUvWithDuv(cct: number, duv: number): [number, number] {
  const uv = planckUv(cct);
  const x = (1_000_000 / cct - 1_000_000 / D65_CCT) / 100;
  const du = polynomial(PLANCK_TANGENT_U, x);
  const dv = polynomial(PLANCK_TANGENT_V, x);
  const length = Math.hypot(du, dv);
  return [uv[0] - (duv * dv) / length, uv[1] + (duv * du) / length];
}

function planckUv(cct: number): [number, number] {
  const x = (1_000_000 / cct - 1_000_000 / D65_CCT) / 100;
  return [polynomial(PLANCK_U, x), polynomial(PLANCK_V, x)];
}

function polynomial(coefficients: readonly number[], x: number): number {
  let result = 0;
  for (let index = coefficients.length - 1; index >= 0; index -= 1) {
    result = result * x + coefficients[index];
  }
  return result;
}

function xyToXyz(xy: readonly [number, number]): [number, number, number] {
  return [xy[0] / xy[1], 1, (1 - xy[0] - xy[1]) / xy[1]];
}

function chromaticAdaptation(
  source: [number, number, number],
  target: [number, number, number],
): [MatrixRow, MatrixRow, MatrixRow] {
  const sourceCone = multiplyVector(BRADFORD, source);
  const targetCone = multiplyVector(BRADFORD, target);
  const scale: Matrix = [
    [targetCone[0] / sourceCone[0], 0, 0],
    [0, targetCone[1] / sourceCone[1], 0],
    [0, 0, targetCone[2] / sourceCone[2]],
  ];
  return multiplyMatrices(multiplyMatrices(BRADFORD_INVERSE, scale), BRADFORD);
}

function multiplyVector(
  matrix: Matrix,
  vector: [number, number, number],
): [number, number, number] {
  return matrix.map(
    (row) => row[0] * vector[0] + row[1] * vector[1] + row[2] * vector[2],
  ) as [number, number, number];
}

function multiplyMatrices(
  left: Matrix,
  right: Matrix,
): [MatrixRow, MatrixRow, MatrixRow] {
  const productRow = (row: number): MatrixRow => [
    left[row][0] * right[0][0] +
      left[row][1] * right[1][0] +
      left[row][2] * right[2][0],
    left[row][0] * right[0][1] +
      left[row][1] * right[1][1] +
      left[row][2] * right[2][1],
    left[row][0] * right[0][2] +
      left[row][1] * right[1][2] +
      left[row][2] * right[2][2],
  ];
  return [productRow(0), productRow(1), productRow(2)];
}

File: crates/lutify-core/src/white_balance.rs
use crate::{LutifyError, Result};

const D65_CCT: f64 = 6504.0;
const D65_XY: [f64; 2] = [0.3127, 0.3290];
const TINT_DUV_PER_STEP: f64 = -0.0005;

// Degree-10 least-squares fits of colour-science's CIE 1931 2° Planckian
// locus over LUTify's complete -100..=100 mired interval. The independent
// Studio fixture constrains the resulting Bradford matrices to 1e-6.
const PLANCK_U: [f64; 11] = [
    0.200_428_082_589_465_35,
    0.021_670_127_838_210_658,
    0.004_941_279_052_065_094,
    -0.000_809_084_031_476_496_4,
    -0.000_113_186_386_143_418_91,
    0.000_106_501_488_138_024_9,
    -0.000_029_459_270_200_989_892,
    -0.000_002_902_435_510_008_949,
    0.000_005_728_170_566_761_21,
    -0.000_001_271_991_161_422_797_4,
    -0.000_000_256_527_231_885_100_6,
];
const PLANCK_V: [f64; 11] = [
    0.310_333_428_902_755_67,
    0.029_893_630_533_631_914,
    -0.004_425_785_338_827_4,
    -0.001_309_960_426_665_987_7,
    0.000_811_335_990_646_606_1,
    -0.000_180_807_490_137_041_64,
    -0.000_019_086_191_775_832_478,
    0.000_035_856_865_207_007_13,
    -0.000_013_826_413_811_352_124,
    -0.000_000_675_784_317_713_839,
    0.000_001_740_341_098_588_157_5,
];
const PLANCK_TANGENT_U: [f64; 11] = [
    0.586_918_218_095_664_9,
    0.289_381_482_664_826_4,
    -0.019_287_578_494_842_157,
    -0.036_230_969_902_795_16,
    0.007_623_792_870_567_206,
    0.003_052_094_086_654_095_3,
    -0.001_040_302_632_497_720_3,
    -0.000_185_296_227_883_120_33,
    0.000_086_661_911_843_005_17,
    0.000_008_512_881_065_854_595,
    -0.000_003_949_015_108_301_125,
];
const PLANCK_TANGENT_V: [f64; 11] = [
    0.809_646_222_310_881_2,
    -0.209_774_657_094_420_28,
    -0.064_908_964_927_516_04,
    0.016_340_113_139_321_086,
    0.008_825_176_667_254_188,
    -0.002_203_072_113_364_394,
    -0.000_994_631_656_435_808_9,
    0.000_305_546_953_852_569_7,
    0.000_076_736_798_465_353_53,
    -0.000_026_349_176_456_116_718,
    -0.000_003_187_068_256_050_521_3,
];

const BRADFORD: [[f64; 3]; 3] = [
    [0.8951, 0.2664, -0.1614],
    [-0.7502, 1.7135, 0.0367],
    [0.0389, -0.0685, 1.0296],
];
const BRADFORD_INVERSE: [[f64; 3]; 3] = [
    [
        0.986_992_905_466_712_3,
        -0.147_054_256_420_990_13,
        0.159_962_651_663_731_25,
    ],
    [
        0.432_305_269_723_394_5,
        0.518_360_271_536_777_6,
        0.049_291_228_212_855_594,
    ],
    [
        -0.008_528_664_575_177_328,
        0.040_042_821_654_084_87,
        0.968_486_695_787_55,
    ],
];
const PROPHOTO_D65_TO_XYZ: [[f64; 3]; 3] = [
    [
        0.755_603_256_421_359,
        0.112_784_921_138_012_72,
        0.082_081_893_435_322_89,
    ],
    [
        0.268_337_925_045_012_8,
        0.715_126_770_695_557_1,
        0.016_535_310_335_320_977,
    ],
    [
        0.003_910_020_350_449_157,
        -0.012_918_708_286_404_542,
        1.097_838_775_355_759_7,
    ],
];
const XYZ_TO_PROPHOTO_D65: [[f64; 3]; 3] = [
    [
        1.403_215_267_115_838_7,
        -0.223_140_097_671_628_46,
        -0.101_553_049_283_432_09,
    ],
    [
        -0.526_271_495_421_098_2,
        1.481_661_091_544_280_5,
        0.017_031_312_123_848_466,
    ],
    [
        -0.011_190_484_846_281_556,
        0.018_230_026_296_336_695,
        0.911_442_663_079_700_9,
    ],
];

/// Relative white-balance recipe applied after the camera's As Shot balance.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct WhiteBalance {
    pub(crate) matrix: [[f32; 3]; 3],
}

impl WhiteBalance {
    /// Creates relative Temperature and Tint adjustments in `-100..=100`.
    /// Zero on both axes is an exact identity transform.
    ///
    /// # Errors
    ///
    /// Returns [`LutifyError::InvalidWhiteBalance`] for non-finite or
    /// out-of-range values.
    pub fn new(temperature: f32, tint: f32) -> Result<Self> {
        if !temperature.is_finite()
            || !tint.is_finite()
            || !(-100.0..=100.0).contains(&temperature)
            || !(-100.0..=100.0).contains(&tint)
        {
            return Err(LutifyError::InvalidWhiteBalance);
        }
        if temperature == 0.0 && tint == 0.0 {
            return Ok(Self::as_shot());
        }

        let target_mired = 1_000_000.0 / D65_CCT + f64::from(temperature);
        let cct = 1_000_000.0 / target_mired;
        let uv = planck_uv_with_duv(cct, f64::from(tint) * TINT_DUV_PER_STEP);
        let denominator = 2.0 * uv[0] - 8.0 * uv[1] + 4.0;
        let target_xy = [3.0 * uv[0] / denominator, 2.0 * uv[1] / denominator];
        let cat = chromatic_adaptation(xy_to_xyz(D65_XY), xy_to_xyz(target_xy));
        let matrix = multiply_matrices(
            &multiply_matrices(&XYZ_TO_PROPHOTO_D65, &cat),
            &PROPHOTO_D65_TO_XYZ,
        );
        // WebGPU and the pixel pipeline consume f32. Matrix construction stays
        // in f64 so the only precision reduction happens at this boundary.
        #[allow(clippy::cast_possible_truncation)]
        let matrix = matrix.map(|row| row.map(|value| value as f32));
        Ok(Self { matrix })
    }

    /// Returns the camera's As Shot balance with no relative correction.
    #[must_use]
    pub const fn as_shot() -> Self {
        Self {
            matrix: [[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]],
        }
    }
}

fn planck_uv_with_duv(cct: f64, duv: f64) -> [f64; 2] {
    let uv = planck_uv(cct);
    let x = ((1_000_000.0 / cct) - 1_000_000.0 / D65_CCT) / 100.0;
    let du = polynomial(&PLANCK_TANGENT_U, x);
    let dv = polynomial(&PLANCK_TANGENT_V, x);
    let length = du.hypot(dv);
    [uv[0] - duv * dv / length, uv[1] + duv * du / length]
}

fn planck_uv(cct: f64) -> [f64; 2] {
    let x = ((1_000_000.0 / cct) - 1_000_000.0 / D65_CCT) / 100.0;
    [polynomial(&PLANCK_U, x), polynomial(&PLANCK_V, x)]
}

fn polynomial(coefficients: &[f64], x: f64) -> f64 {
    coefficients
        .iter()
        .rev()
        .fold(0.0, |result, coefficient| result.mul_add(x, *coefficient))
}

fn xy_to_xyz(xy: [f64; 2]) -> [f64; 3] {
    [xy[0] / xy[1], 1.0, (1.0 - xy[0] - xy[1]) / xy[1]]
}

fn chromatic_adaptation(source: [f64; 3], target: [f64; 3]) -> [[f64; 3]; 3] {
    let source_cone = multiply_vector(&BRADFORD, source);
    let target_cone = multiply_vector(&BRADFORD, target);
    let scale = [
        [target_cone[0] / source_cone[0], 0.0, 0.0],
        [0.0, target_cone[1] / source_cone[1], 0.0],
        [0.0, 0.0, target_cone[2] / source_cone[2]],
    ];
    multiply_matrices(&multiply_matrices(&BRADFORD_INVERSE, &scale), &BRADFORD)
}

fn multiply_vector(matrix: &[[f64; 3]; 3], vector: [f64; 3]) -> [f64; 3] {
    matrix.map(|row| row[0] * vector[0] + row[1] * vector[1] + row[2] * vector[2])
}

fn multiply_matrices(left: &[[f64; 3]; 3], right: &[[f64; 3]; 3]) -> [[f64; 3]; 3] {
    std::array::from_fn(|row| {
        std::array::from_fn(|column| {
            left[row][0] * right[0][column]
                + left[row][1] * right[1][column]
                + left[row][2] * right[2][column]
        })
    })
}

#[cfg(test)]
mod tests {
    use serde::Deserialize;

    use super::*;

    const STUDIO_REFERENCE: &str =
        include_str!("../../../tests/fixtures/studio-white-balance-reference.json");

    #[derive(Deserialize)]
    struct Reference {
        studio_commit: String,
        cases: Vec<Case>,
    }

    #[derive(Deserialize)]
    struct Case {
        temperature: f32,
        tint: f32,
        matrix: [[f64; 3]; 3],
    }

    #[test]
    fn matrices_match_raw_alchemy_studio() {
        let reference: Reference = serde_json::from_str(STUDIO_REFERENCE).unwrap();
        assert_eq!(
            reference.studio_commit,
            "c9823146ba674be52d62f4c55b4c649f796bafd0"
        );
        for case in reference.cases {
            let actual = WhiteBalance::new(case.temperature, case.tint)
                .unwrap()
                .matrix;
            for (row, actual_row) in actual.iter().enumerate() {
                for (column, actual_value) in actual_row.iter().enumerate() {
                    assert!(
                        (f64::from(*actual_value) - case.matrix[row][column]).abs() < 1.0e-6,
                        "({}, {}) [{row}][{column}] actual={} expected={}",
                        case.temperature,
                        case.tint,
                        actual_value,
                        case.matrix[row][column],
                    );
                }
            }
        }
    }

    #[test]
    fn as_shot_is_exact_and_invalid_values_are_rejected() {
        assert_eq!(
            WhiteBalance::new(0.0, 0.0).unwrap(),
            WhiteBalance::as_shot()
        );
        assert_eq!(
            WhiteBalance::new(100.1, 0.0).unwrap_err(),
            LutifyError::InvalidWhiteBalance
        );
        assert_eq!(
            WhiteBalance::new(0.0, f32::NAN).unwrap_err(),
            LutifyError::InvalidWhiteBalance
        );
    }
}
